components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::codec::{
+    mysql::{decimal::DecimalEncoder, json::JsonEncoder},
+    Datum, Error, Result,
+};
+use codec::prelude::*;
+
+use super::{BoundedI64, BoundedU64};
+
+const OFFSET_SIZE: usize = 2;
+const LARGE_OFFSET_SIZE: usize = 4;
+const ID_SIZE: usize = 1;
+const LARGE_ID_SIZE: usize = 4;
+
+// the row format is:
+// | version | flag | number_of_non_null_values | number_of_null_values | non_null_value_ids | null_value_ids | value_offsets | values
+// short spec of each field, check https://github.com/pingcap/tidb/blob/master/docs/design/2018-07-19-row-format.md for details
+// version: 1 byte
+// flag: 1 byte, when there's id greater than 255, value is 1, otherwise 0
+// number of non-null values: 2 bytes
+// number of null values: 2 bytes
+// column ids:  ids of non-null values + ids of null values, when flag == 1 (big), id is 4 bytes, otherwise 1 byte
+// non-null values offset: when big, offset is 4 bytes, otherwise 2 bytes
+pub fn encode(datums: Vec<Datum>, ids: &[i64]) -> Result<Vec<u8>> {
+    let row = Row::new(datums, ids);
+    let mut wtr = Vec::with_capacity(row.approx_size());
+    wtr.write_row(row)?;
+    Ok(wtr)
+}
+
+trait RowEncoder: NumberEncoder {
+    fn write_row(&mut self, row: Row) -> Result<()> {
+        let mut offset_wtr = Vec::with_capacity(row.offsets_len());
+        let mut value_wtr = Vec::with_capacity(row.value_approx_size);
+
+        self.write_u8(super::CODEC_VERSION)?;
+        self.write_flag(row.is_big)?;
+        self.write_u16_le(row.non_null_cols.len() as u16)?;
+        self.write_u16_le(row.null_cols.len() as u16)?;
+
+        for (datum, id) in row.non_null_cols {
+            self.write_id(row.is_big, id)?;
+            value_wtr.write_datum(datum)?;
+            offset_wtr.write_offset(row.is_big, value_wtr.len())?;
+        }
+        for (_, id) in row.null_cols {
+            self.write_id(row.is_big, id)?;
+        }
+        self.write_bytes(&offset_wtr)?;
+        self.write_bytes(&value_wtr)?;
+        Ok(())
+    }
+
+    #[inline]
+    fn write_flag(&mut self, is_big: bool) -> codec::Result<()> {
+        self.write_u8(if is_big { 1 } else { 0 })
+    }
+
+    #[inline]
+    fn write_id(&mut self, is_big: bool, id: i64) -> codec::Result<()> {
+        if is_big {
+            self.write_u32_le(id as u32)
+        } else {
+            self.write_u8(id as u8)
+        }
+    }
+
+    #[inline]
+    fn write_offset(&mut self, is_big: bool, offset: usize) -> codec::Result<()> {
+        if is_big {
+            self.write_u32_le(offset as u32)
+        } else {
+            self.write_u16_le(offset as u16)
+        }
+    }
+}
+impl RowEncoder for Vec<u8> {}
+
+trait DatumEncoder: NumberEncoder + DecimalEncoder + JsonEncoder {
```

Comment:
Prefer to implement it for the `ScalarValue` instead (since we are going to remove `Datum` very soon)

Reply:
If using `ScalarValue`, how should I detect the `Int`  is unsigned or not, by passing a `ColumnInfo`?
I saw somewhere was using `ExprType` of `Expr`

Reply:
Yes! You can pass in a FieldType to know it's signed / unsigned, as well as it's f64 or f32, etc.

Reply:
Changed to `ScalarValue`, PTAL, thanks

Reply:
Do we need to change the `DatumEncoder` to `ScalarValueEncoder`?